src/libmongoc/src/mongoc/mongoc-topology-description.c

```diff
@@ -649,6 +654,57 @@ _mongoc_topology_description_validate_max_staleness (
    return true;
 }
 
+static bool
+_check_any_server_less_than_wire_version_13 (const void *sd_,
+                                             void *any_too_old_)
+{
+   const mongoc_server_description_t *sd = sd_;
+   bool *any_too_old = any_too_old_;
+   if (sd->max_wire_version < 13) {
+      *any_too_old = true;
+   }
+   return true;
+}
+
+/**
+ * @brief Calculate the read mode that we should be using, based on what was
+ * requested and what is available in the topology.
+ *
+ * Per the CRUD spec, if the requested read mode is *not* primary, and *any*
+ * server in the topology has a wire version <13, we must override the read
+ * mode preference with "primary." Wire version 13 indicates support on a
+ * secondary server for using aggregate pipelines that contain writing stages
+ * (i.e. '$out' and '$merge').
+ */
+static mongoc_read_mode_t
+_calc_effective_read_mode (const mongoc_topology_description_t *td,
+                           mongoc_ss_optype_t optype,
+                           mongoc_read_mode_t requested_read_mode)
+{
+   if (requested_read_mode == MONGOC_READ_PRIMARY ||
+       requested_read_mode == MONGOC_READ_PRIMARY_PREFERRED) {
```

Comment:
This is a good question. My intuition was to treat them as equivalent cases, but the literal text says otherwise. @jmikola do you know what the correct behavior would be for 'primaryPreferred'? Is it even applicable to this code path?

Reply:
Sorry for the delayed response. `primary` should definitely be distinguished from `primaryPreferred` when we're concerned with whether to support `$out`/`$merge` on secondaries or not. I'm not entirely sure about the context of this function, but if it's intended to determine fall back behavior then we should consider `primary` separately.

There are ramifications to the read preference beyond what libmongoc itself will select. In the case of a load balanced or sharded cluster, we're potentially forwarding the read preference on to mongos for evaluation.

Reply:
@vector-of-bool I don't believe this has been addressed yet?

Reply:
Darn. I had one last commit that was un-pushed. Pushed now.